myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md

```diff
@@ -0,0 +1,747 @@
+            idata_ar, var_names=["likelihood", "yhat_fut"], predictions=True, random_seed=100
+        )
+
+    return idata_ar, idata_preds, AR
+```
+
+We will fit this model by specifying priors on the negative trend and the range of the standard deviation to respect the direction of the data drift.
+
+```{code-cell} ipython3
+priors_0 = {
+    "coefs": {"mu": [0.2, 0.2], "sigma": [0.5, 0.03], "size": 2},
+    "alpha": {"mu": -4, "sigma": 0.1},
+    "beta": {"mu": -0.1, "sigma": 0.2},
+    "sigma": 8,
+    "init": {"mu": -4, "sigma": 0.1, "size": 1},
+}
+
+
+idata_trend, preds_trend, model = make_latent_AR_trend_model(y_t, priors_0, full_sample=True)
+```
+
+```{code-cell} ipython3
+pm.model_to_graphviz(model)
+```
+
+We can see the structure more clearly with the plate notation, and this additional structure has helped to appropriately fit the directional trend of the timeseries data.
+
+```{code-cell} ipython3
+plot_fits(idata_trend, preds_trend);
+```
+
+```{code-cell} ipython3
+az.summary(idata_trend, var_names=["coefs", "sigma", "alpha", "beta"])
+```
+
+## Complicating the picture further
+
+Next we'll add a seasonal component to our data and see how we can recover this aspect of the data with a bayesian structural timeseries model.
+
+```{code-cell} ipython3
+t_data = list(range(200))
+n_order = 10
+periods = np.array(t_data) / 7
+
+fourier_features = pd.DataFrame(
+    {
+        f"{func}_order_{order}": getattr(np, func)(2 * np.pi * periods * order)
+        for order in range(1, n_order + 1)
+        for func in ("sin", "cos")
+    }
+)
+
+y_t_s = y_t + 20 * fourier_features["sin_order_1"]
+
+fig, ax = plt.subplots(figsize=(10, 4))
+ax.plot(y_t_s)
+ax.set_title("AR + Trend + Seasonality");
+```
+
+The key to fitting this model is to understand that we're now passing in synthetic fourier features to help account for seasonality effects. This works because (roughly speaking) we're trying to fit a complex oscillating phenomena using a weighted combination of sine and cosine waves. So we add these sine waves and consine waves like we would add any other feature variables in a regression model. 
+
+However, since we're using this weighted sum to fit the observed data, the model now expects a linear combination of those synthetic features **also** in the prediction step. As such we need to be able to supply those features even out into the future. This fact remains key for any other type of predictive feature we might want to add e.g. day of the week, holiday flag or any other. If a feature is required to fit the observed data the feature must be available in the prediction step too. 
+
+### Specifying the Trend + Seasonal Model
+
+```{code-cell} ipython3
+def make_latent_AR_trend_seasonal_model(
+    ar_data, ff, priors, prediction_steps=250, full_sample=True, samples=2000
+):
+    with pm.Model() as AR:
+        pass
+
+    ff = ff.to_numpy().T
+    t_data = list(range(len(ar_data)))
+    AR.add_coord("obs_id", t_data, mutable=True)
```

Comment:
Same suggestion about adding coords and dims. As you can see the resulting graphviz is getting quite gnarly, so by adding these it should act to make this slightly easier to visually parse by the reader.

Reply:
This is allot better: 
<img width="1265" alt="image" src="https://user-images.githubusercontent.com/10588167/200558463-6f1ac770-6298-44a3-b258-3062638faf38.png">